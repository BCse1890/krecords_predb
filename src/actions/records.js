import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment';

// ADD RECORD
export const addRecord = (record) => ({
    type: 'ADD_RECORD',
    record
});

export const startAddRecord = (recordData = {}) => {
    return (dispatch) => {
        const {
            name='', 
            address='', 
            contactTel='', 
            email='', 
            relativeName='',
            relativeContactTel='', 
            mainDojo='', 
            dateJoinClub=0, 
            gradeLevel=10, 
            gradingDate=undefined, 
            classesSinceGrading=0
        } = recordData;
        
        const record = {name, mainDojo, gradeLevel, gradingDate, dateJoinClub, 
            address, contactTel, email, relativeName, relativeContactTel, 
            classesSinceGrading };

        database.ref('records').push(record).then((ref) => {
            dispatch(addRecord({
                id: ref.key,
                ...record
            }));
        });
    };
};

/* 
// ADD RECORD
export const addRecord = (record) => ({
    type: 'ADD_RECORD',
    record
});

export const startAddRecord = (recordData = {}) => {
    return (dispatch) =>  {
        const {
            name='', 
            address='', 
            contactTel='', 
            email='', 
            relativeName='',
            relativeContactTel='', 
            mainDojo='', 
            dateJoinClub=0, 
            gradeLevel=10, 
            gradingDate=0, 
            classesSinceGrading=0
        } = recordData;
        
        const record = {name, mainDojo, gradeLevel, gradingDate, dateJoinClub, 
            address, contactTel, email, relativeName, relativeContactTel, 
            classesSinceGrading };

        // return next line so that can use chained promises in test
        // records.test.js    
        database.ref('records').push(record).then((ref) => {
            dispatch(addRecord({
                id: ref.key,
                ...record
            }));
        });
    };
};
 */
// DELETE RECORD
export const deleteRecord = ({ id } = {}) => ({
    type: 'DELETE_RECORD',
    id
});

// EDIT RECORD
export const editRecord = (id, updates) => ({
    type: 'EDIT_RECORD',
    id,
    updates
});