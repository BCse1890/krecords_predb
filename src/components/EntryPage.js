import React from 'react';
import RecordList from './RecordList';
import RecordListFilters from './RecordListFilters';

const EntryPage = () => (
    <div>
        <RecordListFilters />
        <RecordList/>
    </div>
);


export default EntryPage;