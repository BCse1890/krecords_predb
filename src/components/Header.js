import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Student Record Webpage</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Homepage</NavLink>
        <NavLink to="/add" activeClassName="is-active">Create Record</NavLink>
        <NavLink to="/edit" activeClassName="is-active">Edit Record</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        <NavLink to="/contact" activeClassName="is-active">Registration details</NavLink> 
    </header>
);

export default Header;