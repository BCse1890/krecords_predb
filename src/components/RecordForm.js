import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
//import 'react-dates/initialize';



const now = moment();
console.log(now.format('D MMM YYYY'));

export default class RecordForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            name: props.record ? props.record.name : '',
            mainDojo: props.record ? props.record.mainDojo : '',
            gradeLevel: props.record ? props.record.gradeLevel : 10,
            note: props.record ? props.record.note : '',
            gradingDate: props.record ? moment(props.record.gradingDate) : moment(),
            //dateJoinClub: props.record ? moment(props.record.dateJoinClub) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onNameChange = (e) => {
        const name = e.target.value;
        this.setState(() => ({ name }));
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onDojoChange = (e) => {
        const mainDojo = e.target.value;
        this.setState(() => ({ mainDojo }));
    }

    onGradeChange = (e) => {
        const gradeLevel = e.target.value;
        this.setState(() => ({ gradeLevel }));
    }

    onGradingDateChange = (gradingDate) => {
        if(gradingDate) {
            this.setState(() => ({ gradingDate }));
        }
    }

    // onDateJoinChange = (dateJoinClub) => {
    //     if(dateJoinClub) {
    //         this.setState(() => ({ dateJoinClub }));
    //     }
    // }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onSubmit = (e) => {
        e.preventDefault();  // prevents full refresh

        if(!this.state.name || !this.state.gradeLevel || !this.state.mainDojo) {
            this.setState(() => ({ error: 'Please provide Name, Main Dojo and Grade Level'}));
        } else {
            this.setState(() => ({ error: ''}));
            console.log('submitted');
            this.props.onSubmit({
                name: this.state.name,
                mainDojo: this.state.mainDojo,
                gradeLevel: this.state.gradeLevel,
                gradingDate: this.state.gradingDate.valueOf(),
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                      type="text"
                      placeholder="Name"
                      autoFocus
                      value={this.state.name}
                      onChange={this.onNameChange}
                    />
                    <br/>    
                    <input
                      type="text"
                      placeholder="Main Dojo"
                      value={this.state.mainDojo}
                      onChange={this.onDojoChange}
                    />
                    <br/>
                    <input 
                        input="number"
                        placeholder="Grade Level"
                        value={this.state.gradeLevel}
                        onChange={this.onGradeChange}
                    />
                    <br/>
                    <label>
                        Grading date:  
                        <SingleDatePicker 
                            date={this.state.gradingDate}
                            onDateChange={this.onGradingDateChange}
                            focused={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={(day) => false}
                        />
                    </label>
                    <br/>
                    <textarea 
                      placeholder="Add notes"
                      value={this.state.note}
                      onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add record</button>
                </form>
            </div>
        )
    }
}