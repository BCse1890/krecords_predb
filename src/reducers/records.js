
// create kRecords Reducer
const recordsReducerDefault = [];

const recordsReducer = (state = recordsReducerDefault, action) => {
    switch(action.type) {
        case 'ADD_RECORD':
            return [
                //spreads out all existing records in array
                // and adds on a new array through action.record
                ...state,
                action.record
            ];   
        case 'DELETE_RECORD':
            return state.filter(({ id }) => (id !== action.id) );
        case 'EDIT_RECORD':
            return state.map((record) => {
                    if(record.id === action.id) {
                        return {
                            ...record,
                            ...action.updates
                        }
                    } else {
                        return record;
                    }
                });
        default:
            return state;
    }
};

export default recordsReducer;