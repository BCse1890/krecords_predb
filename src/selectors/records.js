import moment from 'moment';

// Get Visible Expenses
export default (records, { text, sortBy, startDate, endDate }) => {
//const getVisibleExpenses = (records, { text, gradingDate, sortBy, startDate, endDate }) => {
    return records.filter((record) => {
        // filter out if startDate not a number or record.created >= startDate;
        const gradeDateMoment = moment(record.gradingDate);
        const startDateMatch = startDate ? startDate.isSameOrBefore(gradeDateMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(gradeDateMoment, 'day') : true;
        
        const textNameMatch = record.name.toLowerCase().includes(text.toLowerCase());
        const textDojoMatch = record.mainDojo.toLowerCase().includes(text.toLowerCase());
        
            
        return startDateMatch && endDateMatch && textDojoMatch || textNameMatch;
    }).sort((a, b) => {
        if(sortBy === 'mainDojo') {
            return a.mainDojo < b.mainDojo ? -1 : 1;
        } else if(sortBy === 'gradeLevel') {
            return a.gradeLevel < b.gradeLevel ? 1 : -1;
        }
    });
};

// export default getVisibleExpenses;