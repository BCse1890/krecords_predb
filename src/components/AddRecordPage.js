import React from 'react';
import { connect } from 'react-redux';
import RecordForm from './RecordForm';
import { startAddRecord } from '../actions/records';

export class AddRecordPage extends React.Component {
    onSubmit = (record) => {
        this.props.startAddRecord(record);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
            <h1>Add Record</h1>
            <RecordForm 
              onSubmit={this.onSubmit}
            />
        </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddRecord: (record) => dispatch(startAddRecord(record))
    });

export default connect(undefined, mapDispatchToProps)(AddRecordPage);