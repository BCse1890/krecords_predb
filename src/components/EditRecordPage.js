import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import RecordForm from './RecordForm';
import { editRecord, deleteRecord } from '../actions/records';

export class EditRecordPage extends React.Component {
    onSubmit = (record) => {
        this.props.editRecord(this.props.record.id, record);
        this.props.history.push('/');
    };
    onDelete = () => {
        console.log("Delete key pressed");
        this.props.deleteRecord({ id: this.props.record.id });
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <RecordForm 
                    record={this.props.record}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onDelete}>Delete</button>
                <br/>
            </div>
        );
    }
};

const mapStateToProps = (state, props) => ({
    record: state.records.find((record) => 
        record.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    editRecord: (id, record) =>  dispatch(editRecord(id, record)),
    deleteRecord: (data) => dispatch(deleteRecord(data))
});

export default connect (mapStateToProps, mapDispatchToProps)(EditRecordPage);