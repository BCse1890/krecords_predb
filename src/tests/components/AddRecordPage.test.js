import React from 'react';
import { shallow } from 'enzyme';
import '../setUpTests';
import { AddRecordPage } from '../../components/AddRecordPage';
import { RecordForm } from '../../components/RecordForm';
import records from '../fixtures/records';

// using spies below
let startAddRecord, history, wrapper;
beforeEach(() => {
    startAddRecord = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddRecordPage 
        startAddRecord={startAddRecord} history={history} />);
});

test('should render AddRecordPage correctly', () => {
    // const onSubmit = jest.fn();
    // const history = { push: jest.fn() };
    // const wrapper = shallow(<AddRecordPage 
    //     onSubmit={onSubmit} history={history} />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    // const onSubmit = jest.fn();
    // const history = { push: jest.fn() };
    // const wrapper = shallow(<AddRecordPage 
    //     onSubmit={onSubmit} history={history} />);
    wrapper.find('RecordForm').prop('onSubmit')(records[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startAddRecord).toHaveBeenLastCalledWith(records[1]);
});
