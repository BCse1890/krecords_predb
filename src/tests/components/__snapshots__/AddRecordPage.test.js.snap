// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddRecordPage correctly 1`] = `
<div>
  <h1>
    Add Record
  </h1>
  <RecordForm
    onSubmit={[Function]}
  />
</div>
`;
